### PR TITLE
syslog/ramlog: improve ramlog performance

### PR DESCRIPTION
**Pull request by anchao**
## Summary

syslog/ramlog: improve ramlog performance

replace char copy to memcpy to improve the performance

## Impact

N/A

## Testing

ci-check

### CODE REVIEW

File: drivers/syslog/ramlog.c [hunk 1]
```diff
@@ -236,47 +235,93 @@ static void ramlog_pollnotify(FAR struct ramlog_dev_s *priv)
 }
 
 /****************************************************************************
- * Name: ramlog_addchar
+ * Name: ramlog_bufcpy
  ****************************************************************************/
 
-static void ramlog_addchar(FAR struct ramlog_dev_s *priv, char ch)
+static void ramlog_bufcpy(FAR struct ramlog_dev_s *priv,
```
Comment (xiaoxiang781216) [inline drivers/syslog/ramlog.c:241]: ramlog_copybuf?

File: drivers/syslog/ramlog.c [hunk 2]
```diff
@@ -355,19 +388,19 @@ static ssize_t ramlog_file_read(FAR struct file *filep, FAR char *buffer,
...
+      /* Determine whether the read pointer is overwritten */
 
-  tail = upriv->rl_tail % priv->rl_bufsize;
+      if (header->rl_head - upriv->rl_tail > priv->rl_bufsize)
+        {
+          upriv->rl_tail = header->rl_head - priv->rl_bufsize;
+        }
 
-  /* Loop until something is read */
+      pos = upriv->rl_tail % priv->rl_bufsize;
```
[the remark this answers is not in the document]
Comment (anchao) [reply drivers/syslog/ramlog.c:402]: Done

File: drivers/syslog/ramlog.c [hunk 3]
```diff
@@ -427,20 +460,27 @@ static ssize_t ramlog_file_read(FAR struct file *filep, FAR char *buffer,
            * tail index.
            */
 
-          ch = header->rl_buffer[tail];
-
-          /* Increment the tail index. */
+          uint32_t ncopy = (header->rl_head % priv->rl_bufsize) - pos;
```
[the remark this answers is not in the document]
Comment (anchao) [reply drivers/syslog/ramlog.c:463]: Done

File: drivers/syslog/ramlog.c (hunk 1 above)
Comment (anchao) [reply drivers/syslog/ramlog.c:241]: done

File: drivers/syslog/ramlog.c [hunk 4]
```diff
@@ -689,13 +729,14 @@ int ramlog_putc(FAR struct syslog_channel_s *channel, int ch)
...
   irqstate_t flags;
+  char cch = ch;
 
   UNUSED(channel);
 
   /* Add the character to the RAMLOG */
 
   flags = enter_critical_section();
-  ramlog_addchar(priv, ch);
+  ramlog_addchars(priv, &cch, 1);
```
[the remark this answers is not in the document]
Comment (anchao) [reply drivers/syslog/ramlog.c:739]: Done

File: drivers/syslog/ramlog.c [hunk 5]
```diff
@@ -236,47 +235,93 @@ static void ramlog_pollnotify(FAR struct ramlog_dev_s *priv)
...
-  /* Pre-pend a carriage before a linefeed */
+  header->rl_head += len;
+}
+
+/****************************************************************************
+ * Name: ramlog_addchars
+ ****************************************************************************/
 
-  if (ch == '\n')
+static void ramlog_addchars(FAR struct ramlog_dev_s *priv,
```
[the remark this answers is not in the document]
Comment (anchao) [reply drivers/syslog/ramlog.c]: Done

File: drivers/syslog/ramlog.c [hunk 6]
```diff
@@ -287,28 +274,64 @@ static void ramlog_addchar(FAR struct ramlog_dev_s *priv, char ch)
...
+  if (len > priv->rl_bufsize)
     {
-      /* Get the next character to output */
+      buffer += (len - priv->rl_bufsize);
+      len = priv->rl_bufsize;
+    }
+
+#ifdef CONFIG_RAMLOG_CRLF
+  FAR char *rlbuf = header->rl_buffer;
+  FAR char *start = (FAR char *)buffer;
```
Comment (xiaoxiang781216) [inline drivers/syslog/ramlog.c:302]: remove, let reuse `buffer`

File: drivers/syslog/ramlog.c [hunk 7]
```diff
@@ -287,28 +274,64 @@ static void ramlog_addchar(FAR struct ramlog_dev_s *priv, char ch)
...
+
+  if (len > priv->rl_bufsize)
     {
-      /* Get the next character to output */
+      buffer += (len - priv->rl_bufsize);
+      len = priv->rl_bufsize;
+    }
+
+#ifdef CONFIG_RAMLOG_CRLF
+  FAR char *rlbuf = header->rl_buffer;
```
Comment (xiaoxiang781216) [inline drivers/syslog/ramlog.c:301]:
```suggestion
  FAR char *buf = header->rl_buffer;
```
move the declaration to the begin of block.

File: drivers/syslog/ramlog.c [hunk 8]
```diff
@@ -236,48 +235,36 @@ static void ramlog_pollnotify(FAR struct ramlog_dev_s *priv)
...
- * Name: ramlog_addchar
+ * Name: ramlog_copybuf
  ****************************************************************************/
 
-static void ramlog_addchar(FAR struct ramlog_dev_s *priv, char ch)
+static void ramlog_copybuf(FAR struct ramlog_dev_s *priv,
+                          FAR const char *buffer, size_t len)
 {
   FAR struct ramlog_header_s *header = priv->rl_header;
+  FAR char *rlbuf = header->rl_buffer;
```
[the remark this answers is not in the document]
Comment (anchao) [reply drivers/syslog/ramlog.c:245]: Done

File: drivers/syslog/ramlog.c [hunk 9]
```diff
@@ -287,28 +274,64 @@ static void ramlog_addchar(FAR struct ramlog_dev_s *priv, char ch)
...
+      memset(header, 0, sizeof(g_sysbuffer));
+      header->rl_magic = RAMLOG_MAGIC_NUMBER;
+    }
+#endif
+
+  if (len > priv->rl_bufsize)
     {
-      /* Get the next character to output */
+      buffer += (len - priv->rl_bufsize);
+      len = priv->rl_bufsize;
```
[the remark this answers is not in the document]
Comment (anchao) [reply drivers/syslog/ramlog.c:297]: Done

File: drivers/syslog/ramlog.c (hunk 7 above)
Comment (anchao) [reply drivers/syslog/ramlog.c:301]: Done

File: drivers/syslog/ramlog.c (hunk 6 above)
Comment (anchao) [reply drivers/syslog/ramlog.c:302]: Done

File: drivers/syslog/ramlog.c [hunk 10]
```diff
@@ -287,28 +274,69 @@ static void ramlog_addchar(FAR struct ramlog_dev_s *priv, char ch)
...
+  FAR struct ramlog_header_s *header = priv->rl_header;
+#endif
+  size_t buflen = len;
   irqstate_t flags;
-  size_t nwritten;
-  char ch;
+#ifdef CONFIG_RAMLOG_CRLF
+  FAR const char *end;
+  FAR const char *pos;
+  FAR char *rlbuf;
```
Comment (xiaoxiang781216) [inline drivers/syslog/ramlog.c:285]:
```suggestion
  FAR char *buf;
```

File: drivers/syslog/ramlog.c [hunk 11]
```diff
@@ -287,28 +274,69 @@ static void ramlog_addchar(FAR struct ramlog_dev_s *priv, char ch)
...
+    {
+      memset(header, 0, sizeof(g_sysbuffer));
+      header->rl_magic = RAMLOG_MAGIC_NUMBER;
+    }
+#endif
+
+  if (buflen > priv->rl_bufsize)
     {
-      /* Get the next character to output */
+      buffer += (buflen - priv->rl_bufsize);
```
Comment (xiaoxiang781216) [inline drivers/syslog/ramlog.c:302]:
```suggestion
      buffer += buflen - priv->rl_bufsize;
```

**Review by xiaoxiang781216 — APPROVED** [review]

File: drivers/syslog/ramlog.c [hunk 12]
```diff
@@ -423,24 +440,39 @@ static ssize_t ramlog_file_read(FAR struct file *filep, FAR char *buffer,
...
+            }
+
           /* The circular buffer is not empty, get the next byte from the
            * tail index.
            */
 
-          ch = header->rl_buffer[tail];
-
-          /* Increment the tail index. */
+          uint32_t pos = upriv->rl_tail % priv->rl_bufsize;
```
Comment (xiaoxiang781216) [inline drivers/syslog/ramlog.c:454]: need move before line 443

File: drivers/syslog/ramlog.c [hunk 13]
```diff
@@ -287,28 +274,69 @@ static void ramlog_addchar(FAR struct ramlog_dev_s *priv, char ch)
...
+        {
+          ramlog_copybuf(priv, buffer, pos - buffer);
+          buffer = pos + 1;
+        }
+
+      /* Pre-pend a carriage before a linefeed */
+
+      else if (*pos == '\n')
+        {
+          buf[header->rl_head++ % priv->rl_bufsize] = '\r';
```
Comment (xiaoxiang781216) [inline drivers/syslog/ramlog.c:325]: should we need ramlog_copybuf here too?
```
if (*pos == '\r' || *pos == '\n')
  {
    ramlog_copybuf(priv, buffer, pos - buffer);
    buffer = pos + 1;
  }

if (*pos == '\n')
  {
    buf[header->rl_head++ % priv->rl_bufsize] = '\r';
    buf[header->rl_head++ % priv->rl_bufsize] = '\n';
  }
```

File: drivers/syslog/ramlog.c [hunk 14]
```diff
@@ -287,28 +274,69 @@ static void ramlog_addchar(FAR struct ramlog_dev_s *priv, char ch)
...
+          buf[header->rl_head++ % priv->rl_bufsize] = '\n';
+          buffer = pos + 1;
+        }
     }
+  while (pos++ != end);
```
Comment (xiaoxiang781216) [inline drivers/syslog/ramlog.c:330]:
```suggestion
  while (++pos != end);
```

Comment (anchao) [reply drivers/syslog/ramlog.c]: Done

File: drivers/syslog/ramlog.c (hunk 12 above)
Comment (anchao) [reply drivers/syslog/ramlog.c:453]: upriv->rl_tail will update on line 447

File: drivers/syslog/ramlog.c (hunk 13 above)
Comment (anchao) [reply drivers/syslog/ramlog.c:325]: Done

File: drivers/syslog/ramlog.c (hunk 12 above)
Comment (xiaoxiang781216) [reply drivers/syslog/ramlog.c:453]: if so, we need separate the declaration and assignment.

Comment (anchao) [reply drivers/syslog/ramlog.c:453]: Done

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
ci error is fixed here: https://github.com/apache/nuttx/pull/11414

**Pull request merged**